### DISCUSSION TITLE
'Undo' a user-specified number of actions, and other QOL improvements

### DISCUSSION THREAD

**Discussion by OPStellar** [OpenRCT2/OpenRCT2#15874] [Feature requests]
Probably requires new save format.

Given the isometric perspective, distance and height can be easily mistaken. Changes in landscaping can cause a tile to become buried to the point of inaccessibility, and deleting things in the tile inspector (cheat) can be frustrating when you accidentally delete something you don't intend to and have to place it down again. Correcting simple errors caused due to perspective costs a lot of in-game money and in-real-life time. I can afford the former, but the latter is just frustrating and makes me want to rage-quit sometimes.

Aspects to consider:
An 'action' would be defined as all changes made from the time a key or the left mouse-button is pressed to the time it is released.
Undoing would not rewrite history; it would merely perform the reverse of any string of actions taken, either for the same cost as was spent to perform the action or for the resale value of deleting instead of un-building a placed object.
There would be no 'Redo' button unless the user specifies a value of '1' undo, in which case it would operate like Photoshop's undo function that is used for making comparisons between two changes. One should only undo for as many mistakes as they know they made.
It would be fine if the 'undo' function is limited only to undoing deletions (as opposed to that on top of undoing other actions like construction) and only as long as the change can be made (i.e. a non-circuitous ride would need to remain closed or else the memory of the last action would be forgotten).
'User-specified number' assumes that there would need to be a limit based on disk space. Some changes (especially those involving RNG) might demand a lot of memory and cache.

**Comment by ocalhoun6** [comment]
Already been suggested in #2256 (And I think you're kind of overcomplicating it? Just a single undo button would be excellent. If you want to undo multiple actions, you just click the button multiple times.)

But maybe it's worth revisiting?

While I _do_ find the "It's really difficult to implement" objection convincing, the "it makes the game too easy" and "OpenRCT2 is not a design tool, it is a theme park simulator" objections don't seem to hold water for me. You know what makes the game *really* easy? Having a simple cheat button for 'win scenario' -- and we've got that. And while OpenRCT2 is *supposed* to be a theme park simulator, a lot of people use it an awful lot more like a design tool when building intricately aesthetic show parks. (And to be clear, the undo button should only undo player actions, *not* turn back time. The undo button wouldn't undo a coaster crashing ... but it could undo placing track piece that accidentally destroys your intricate scenery.)

It's not a feature I'd expect to see anytime soon (because it *would* be really difficult to implement), but it _is_ something I think should be on the to-do list for future versions. I think we've all been there at some point, especially when it comes to accidentally destroying scenery.

**Comment by duncanspumpkin** [comment]
I've already made a prototype of undo and it's something we are slowly working towards. I want to rework how the code that handles removing scenery in the way (map clearance) before taking the prototype any further. This isn't related to nsf. I suspect the way it will eventually work is that it will only undo what can be performed without failure or depending on another action.

**Comment by TCMOREIRA** [comment]
Hi! I came here to ask the same thing! I find it really frustrating that this game doesn't have undo/redo buttons! I tried introducing this game to my Girlfriend (who usually loves this kind of game and somehow had missed playing this) and the first thing she complained about was "How's it possible that there's no undo button?!".

Has there been any progress on this?
How can I help?
